### PR TITLE
Improve text field input

### PR DESCRIPTION
**Pull request by W1W1-M**
Hi @sindresorhus 
This PR is a first draft for issue #3 [Validate text fields ](https://github.com/sindresorhus/System-Color-Picker/issues/3)

I first tried a border color for the hex text field but it interfered with blue border when the text field is focused. Instead I added a SF symbol image that switches depending if the field is valid.

Capture d’écran 2022-08-17 à 14 32 19
Capture d’écran 2022-08-17 à 14 36 32

The implemented colorChecker class uses ObservedObject and the onChange modifier to run on user input, it also trims unwanted characters.

The code still needs to be implemented for hsl, rgb & lch text inputs. Also needs cleaning, optimizing & testing but I wanted an initial 👍 before going ahead.

Regards

PS : I have more experience with iOS app development, this is helping me get to grips with Mac app development 😄

### CODE REVIEW

**Comment by sindresorhus** [comment]
Thanks for working on this. I don't think this is the right approach though.

Here's what I'm thinking:

1. Abstract all of the `hexColorView` type views into a single (general) view that can be reused for each of the colors.
2. The validation logic is already implemented, if `let newColor = NSColor(hexString: hexColor.trimmingCharacters(in: .whitespaces))` succeeds, it means the color is valid.

I'm also not a big fan of the symbols. We don't need a success indication, just a failure one. So I think just coloring the text red if it's invalid is enough.

File: Color Picker/ColorPickerScreen.swift [hunk 1]
```diff
@@ -156,45 +157,47 @@ struct ColorPickerScreen: View {
...
 					}
-
 					if
 						isTextFieldFocusedHex,
 						!isPreventingUpdate,
 						let newColor = NSColor(hexString: hexColor.trimmingCharacters(in: .whitespaces))
 					{
 						colorPanel.color = newColor
 					}
-
```
Comment (sindresorhus) [inline Color Picker/ColorPickerScreen.swift:183]: Don't make unrelated changes.

**Comment by W1W1-M** [comment]
> 1. Abstract all of the `hexColorView` type views into a single (general) view that can be reused for each of the color inputs.

Ok, I'll work on that and update accordingly
> 2. The validation logic is already implemented, if `let newColor = NSColor(hexString: hexColor.trimmingCharacters(in: .whitespaces))` succeeds, it means the color is valid. If not, you can show the invalid state.

I didn't realize all the logic was already there 😅  Thanks for pointing this out !
> I'm also not a big fan of the symbols. We don't need a success indication, just a failure one. So I think just coloring the text red if it's invalid is enough.

Ok, simple and straightforward is good 😄 I'll update accordingly

**Comment by W1W1-M** [comment]
So this is still WIP, but some feedback to know if this is more inline with the proposed approach would be nice 😃 

As for hex codes: they are limited to a small set of characters and also in length, I've added some code to filter and trim hex inputs (ex: #123456, FFFFFF, etc). This helps users input only valid hex codes but also makes the input rigid and removes the need to indicate (in red) an invalid input. Maybe a compromise of just trimming excess characters without filtering is a good balance ?

File: Color Picker/ColorPickerScreen.swift [hunk 2]
```diff
@@ -127,6 +127,126 @@ private struct BarView: View {
...
+struct ColorInputView: View {
+    @EnvironmentObject private var appState: AppState
+    @State private var textColor: Color = .primary
+    @Binding var inputColorText: String
+    @Binding var isTextFieldFocused: Bool
+    @Binding var isPreventingUpdate: Bool
+    let colorPanel: NSColorPanel
+    let textFieldFontSize: Double
+    let inputColorType: colorType
+    enum colorType: String, CaseIterable {
```
Comment (sindresorhus) [inline Color Picker/ColorPickerScreen.swift:139]: This already exists: https://github.com/sindresorhus/System-Color-Picker/blob/2410aba60bc4d6f0194654f23865c485244e4761/Color%20Picker/Constants.swift#L72

File: Color Picker/ColorPickerScreen.swift [hunk 3]
```diff
@@ -127,6 +127,126 @@ private struct BarView: View {
...
+        case hex = "Hex"
+        case hsl = "HSL"
+        case rgb = "RGB"
+        case lch = "LCH"
+    }
+    var colorKeyboardShortcut: KeyEquivalent {
```
Comment (sindresorhus) [inline Color Picker/ColorPickerScreen.swift:145]: This should be on the `ColorFormat` enum

**Comment by sindresorhus** [comment]
> So this is still WIP, but some feedback to know if this is more inline with the proposed approach would be nice 😃

Yes, it's the correct direction.

**Comment by sindresorhus** [comment]
> This helps users input only valid hex codes but also makes the input rigid

Trimming whitespace is ok, but I don't think we should prevent some characters. Users will not understand that it's because it's not valid Hex. I think it's better to just make it red in that case.

File: Color Picker/ColorPickerScreen.swift (hunk 2 above)
Comment (W1W1-M) [reply Color Picker/ColorPickerScreen.swift]: OK

File: Color Picker/ColorPickerScreen.swift (hunk 3 above)
Comment (W1W1-M) [reply Color Picker/ColorPickerScreen.swift]: I've moved keyboardShortcut into an extension of ColorFormat within ColorPickerScreen. 

Would it be better to centralise this extension in Constants as not to have class extensions in different files ? or is it better to keep here, in order to limit access control of extension to this file only ?

**Comment by W1W1-M** [comment]
On user color input, `updateColorsFromPanel` method needs to be called to update other color codes. This is currently in `ColorPickerScreen`. In order to be able to call it from `ColorInputView` I'm thinking it could be moved to `AppState` `EnvironmentObject` along with required properties (as `@Published` properties) ? 
Or maybe there is a smarter way to call it ? 
Or does it need a bigger rethink as the comment indicates: 

> TODO: Find a better way to handle this.

File: Color Picker/AppState.swift [hunk 4]
```diff
@@ -128,6 +130,14 @@ final class AppState: ObservableObject {
 		}
 	}
 
+    @Published var hexColor = ""
+    @Published var hslColor = ""
+    @Published var rgbColor = ""
+    @Published var lchColor = ""
+    @Published var isPreventingUpdate = false
```
Comment (sindresorhus) [inline Color Picker/AppState.swift:137]: Why are these moved to AppState? AppState is for state that is required globally. This state is only needed for picker screen. I think you can just pass `updateColorsFromPanel` to the `ColorPickerScreen` view.

File: Color Picker/ColorPickerScreen.swift [hunk 5]
```diff
@@ -127,22 +127,149 @@ private struct BarView: View {
 	}
 }
 
+struct ColorInputView: View {
+    @EnvironmentObject private var appState: AppState
+    @State private var textColor: Color = .primary
```
Comment (sindresorhus) [inline Color Picker/ColorPickerScreen.swift:132]: It's better to use an optional and default to `nil`.

File: Color Picker/ColorPickerScreen.swift [hunk 6]
```diff
@@ -127,22 +127,149 @@ private struct BarView: View {
...
+            // TODO: When I use `TextField`, add the copy button using `.safeAreaInset()`.
+            NativeTextField(
+                text: $inputColorText,
+                placeholder: inputColorType.rawValue,
+                font: .monospacedSystemFont(ofSize: textFieldFontSize, weight: .regular),
+                isFocused: $isTextFieldFocused,
+                textColor: textColor
+            )
+                .controlSize(.large)
+                .onChange(of: inputColorText) {
```
Comment (sindresorhus) [inline Color Picker/ColorPickerScreen.swift:150]: Move the body out into a private function on the view. It makes it easier to see just the view logic.

File: Color Picker/ColorPickerScreen.swift [hunk 7]
```diff
@@ -127,22 +127,149 @@ private struct BarView: View {
...
+
+                        if NSColor(cssLCHString: inputColorText.trimmingCharacters(in: .whitespaces)) != nil {
+                            textColor = .primary
+                        } else {
+                            textColor = .red
+                        }
+
+                        if !appState.isPreventingUpdate {
+                            appState.updateColorsFromPanel(excludeLCH: true, preventUpdate: true, color: colorPanel.color)
+                        }
```
Comment (sindresorhus) [inline Color Picker/ColorPickerScreen.swift:238]: There's a lot of duplicated logic in each case here. Would be good to reduce that.

For example, `appState.updateColorsFromPanel` could be updated to accept an enum instead of `excludeLCH: true` type parameters.

File: Color Picker/ColorPickerScreen.swift [hunk 8]
```diff
@@ -155,222 +282,83 @@ struct ColorPickerScreen: View {
...
+                    textFieldFontSize: textFieldFontSize,
+                    inputColorType: .rgb
+                )
 			}
 			if shownColorFormats.contains(.lch) {
-				lchColorView
+                ColorInputView(
+                    inputColorText: $appState.lchColor,
+                    isTextFieldFocused: $isTextFieldFocusedLCH,
+                    colorPanel: colorPanel,
```
Comment (sindresorhus) [inline Color Picker/ColorPickerScreen.swift:319]: The color panel is available on the AppState, so you can access it from there instead.

File: Color Picker/ColorPickerScreen.swift [hunk 9]
```diff
@@ -155,222 +282,83 @@ struct ColorPickerScreen: View {
...
+                    textFieldFontSize: textFieldFontSize,
+                    inputColorType: .lch
```
Comment (sindresorhus) [inline Color Picker/ColorPickerScreen.swift:321]: This parameter should be first.

File: Color Picker/ColorPickerScreen.swift [hunk 10]
```diff
@@ -155,222 +282,83 @@ struct ColorPickerScreen: View {
...
-				.buttonStyle(.borderless)
-				.contentShape(.rectangle)
-				.keyboardShortcut("l", modifiers: [.shift, .command])
-		}
-	}
-
 	var body: some View {
 		VStack {
 			BarView()
 			if shownColorFormats.contains(.hex) {
```
Comment (sindresorhus) [inline Color Picker/ColorPickerScreen.swift:288]: You can just loop over the enum instead of manually defining each view. That was the point of extracting it into a reusable view.

File: Color Picker/Utilities.swift [hunk 11]
```diff
@@ -1021,6 +1021,7 @@ struct NativeTextField: NSViewRepresentable {
 	var isFirstResponder = false
 	@Binding var isFocused: Bool // Note: This is only readable.
 	var isSingleLine = true
+    var textColor: Color
```
Comment (sindresorhus) [inline Color Picker/Utilities.swift:1024]: Nitpick: Should be below `font`.

File: Color Picker/ColorPickerScreen.swift [hunk 12]
```diff
@@ -127,22 +127,149 @@ private struct BarView: View {
...
 	@Default(.shownColorFormats) private var shownColorFormats
 	@Default(.largerText) private var largerText
-	@State private var hexColor = ""
-	@State private var hslColor = ""
-	@State private var rgbColor = ""
-	@State private var lchColor = ""
 	@State private var isTextFieldFocusedHex = false
 	@State private var isTextFieldFocusedHSL = false
 	@State private var isTextFieldFocusedRGB = false
 	@State private var isTextFieldFocusedLCH = false
```
Comment (sindresorhus) [inline Color Picker/ColorPickerScreen.swift:272]: This could be a single state that accepts an optional `ColorFormat`.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

**Comment by sindresorhus** [comment]
Closing for lack of activity.
